salesagility/SuiteCRM
# FIX #3005 - New cache system cache lite

**Expected Behavior**

A new caching method could be implemented using the CacheLite library[1] to prevent many queries being executed to the database, and as a new cache method for SuiteCRM.

[1] https://pear.php.net/package/Cache_Lite/

**Actual Behavior**

As we know SuiteCRM has many cache classes but use by default SugarCacheAbstract class. This cache method use RAM. This is good because RAM is fast, but the cache lasts for a limited time: the request, so in every request has to regenerate the cache. Besides the cache is not shared between users.

CacheLite stores the cache in files, which is slower than using RAM, but has many other benefices:

    

- The cache last much more time than the request, and you can configure that duration ($sugar_config['CacheLite']['LifeTime'] = 600; in config_override.php)

- The cache is shared between users.

-  Reseting the cache is as simple as deleting the cache files.
    
- You can separate the cache in 'groups', so you can delete the cache of certain groups and the whole.


## Motivation and Context

Prevent many queries being executed to the database

## How To Test This

Go to Admin, and click in Repair:


![suitecrm_admin1](https://cloud.githubusercontent.com/assets/939888/22445135/c9b8f5ac-e724-11e6-8e97-70ede7da64b2.png)

Now we have a new option "Flush the cache". After click in that option, the cache is clean:

![suitecrm_admin2](https://cloud.githubusercontent.com/assets/939888/22445190/0bb64c02-e725-11e6-9b84-3464d3548ed5.png)

You can see the cache actions in suitecrm.log

![suitecrm_log](https://cloud.githubusercontent.com/assets/939888/22445221/2f34ab1a-e725-11e6-9062-4aa4bfa28cf5.png)


The cache files are stored in cache/ folder

![suitecrm_cache_files](https://cloud.githubusercontent.com/assets/939888/22445239/48ad87d8-e725-11e6-9a1b-9cd1277d1697.png)



**Some notes about this:**

- Now the default Cache method is the one we implement. This can be changed in the file: include/SugarCache/SugarCachePHPCacheLite.php
May be this could be seted up in config_override.php.

- We saw that the queries executed by the class OutboundEmail (include/OutboundEmail/OutboundEmail.php) were a lot and always the same, and changes rarely. This was the first implementation of the new cache method.
Example: In 53 minutes (the duration of a 200MB SuiteCRM log before it rotates) this class executed 100.000 queries(!) of a total of 413.000 queries.

- We made a test in a development instance of our SuiteCRM implementation, and saw that the queries executed were 44% less!.

This is a good performance improvement for SuiteCRM, and we have it in production.





## Types of changes


### Final checklist
